### PR TITLE
Feature requests: Switch now optional for service calls | New default icons | Reload `.yaml` w/o restart | `adapt_delay` ms

### PR DESCRIPTION
**Pull request by th3w1zard1**
This is a collaborative of all the small kind/feature enhancements I've found scowering the issues board. Until a merge happens, I'll continually update this with smaller feature requests.

Current features and their origin thread:

- #210
- #233

all service calls no longer require an entity_id, you can now pass a light and the integration is smart enough to find the switch for you. This is probably the biggest feature, services.yaml has been updated to streamline the user experience tenfold, as it'll show selectors in the GUI when using services in scripts/automations.
Huge shoutout to @firstof9 and @Raman325 in the HASS discord for helping me figure this out.

- #274 icons are implemented.
- #104 

You can now reload any changes to the yaml file without restarting your home assistant. Should show a 'reload' button in your integrations page or you can call the homeassistant reload integration service call.
See https://community.home-assistant.io/t/how-to-allow-custom-compontent-for-yaml-configuration-reloading/391190/4

Please feel free to clean up any of the code. I apologize for the numerous small commits I heavily relied on `pre-commit` to find syntax errors before each test.

### CODE REVIEW

**Comment by firstof9** [comment]
Looks like you may need to update your tests.

File: custom_components/adaptive_lighting/switch.py [hunk 1]
```diff
@@ -66,15 +64,14 @@
     SUN_EVENT_SUNRISE,
     SUN_EVENT_SUNSET,
 )
-from homeassistant.core import (
+from homeassistant.core import (  # ServiceCall,
     Context,
     Event,
     HomeAssistant,
-    ServiceCall,
```
Comment (firstof9) [inline custom_components/adaptive_lighting/switch.py:73]: Keep `ServiceCall` imported

File: custom_components/adaptive_lighting/switch.py [hunk 2]
```diff
@@ -237,57 +235,106 @@ def _split_service_data(service_data, adapt_brightness, adapt_color):
...
-async def handle_apply(switch: AdaptiveSwitch, service_call: ServiceCall):
-    """Handle the entity service apply."""
-    hass = switch.hass
-    data = service_call.data
-    all_lights = data[CONF_LIGHTS]
-    if not all_lights:
-        all_lights = switch._lights
-    all_lights = _expand_light_groups(hass, all_lights)
-    switch.turn_on_off_listener.lights.update(all_lights)
+def parseServiceArgs(hass: HomeAssistant, service_call):
```
[the remark this answers is not in the document]
Comment (firstof9) [reply custom_components/adaptive_lighting/switch.py:238]: Might as well type both. No reason not to.

File: custom_components/adaptive_lighting/switch.py [hunk 3]
```diff
@@ -342,33 +397,98 @@ async def async_setup_entry(
         update_before_add=True,
     )
 
+    @callback
+    async def handle_apply(service_call):
```
Comment (firstof9) [inline custom_components/adaptive_lighting/switch.py:401]:
```suggestion
    async def handle_apply(service_call: ServiceCall):
```

File: custom_components/adaptive_lighting/switch.py [hunk 4]
```diff
@@ -342,33 +397,98 @@ async def async_setup_entry(
...
+                    data[ATTR_ADAPT_COLOR],
+                    data[CONF_PREFER_RGB_COLOR],
+                    force=True,
+                    context=thisSwitch.create_context(
+                        "service", parent=service_call.context
+                    ),
+                )
+
+    @callback
+    async def handle_set_manual_control(service_call):
```
Comment (firstof9) [inline custom_components/adaptive_lighting/switch.py:427]:
```suggestion
    async def handle_set_manual_control(service_call: ServiceCall):
```

File: .github/workflows/automerge.yaml [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+on:
+  push:
+    branches:
+      - '!master'
+      - '!dev-branch'
+
+env:
+  # path to the branch you're automerging into.
+  AUTOMERGE_INTO: https://github.com/th3w1zard1/adaptive-lighting/tree/dev-branch.git
```
Comment (basnijholt) [inline .github/workflows/automerge.yaml:11]: This is your fork

Comment (th3w1zard1) [reply .github/workflows/automerge.yaml]: My mistake keeping my workflow separate from forks is a problem I haven't managed to fix yet. Removed in the last commit `708677e`

File: custom_components/adaptive_lighting/switch.py (hunk 4 above)
Comment (th3w1zard1) [reply custom_components/adaptive_lighting/switch.py]: done!

File: custom_components/adaptive_lighting/switch.py [hunk 6]
```diff
@@ -238,57 +237,136 @@ def _split_service_data(service_data, adapt_brightness, adapt_color):
...
+                service_call.data,
+            )
+            raise ValueError(
+                "adaptive-lighting: Multiple switches passed with lights argument"
             )
+        switch = []
+        for unique_id in switch_unique_id:
+            ent_reg = entity_registry.async_get(hass)
+            ent_entry = ent_reg.async_get(unique_id)
+            # `AttributeError: 'NoneType' object has no attribute 'config_entry_id'` fix later maybe
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:281]: When does this exception occur?

File: custom_components/adaptive_lighting/switch.py [hunk 7]
```diff
@@ -238,57 +237,136 @@ def _split_service_data(service_data, adapt_brightness, adapt_color):
...
+            _LOGGER.debug(
+                "_parse_service_args: switch: %s, ent_reg: %s, ent_entry: %s, config_id: %s",
+                switch,
+                ent_reg,
+                ent_entry,
+                config_id,
+            )
+            switch.append(hass.data[DOMAIN][config_id]["instance"])
+    elif lights:
+        # all_adapt_switches = integration_entities(hass, DOMAIN)
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:292]: Remove this commented code

File: custom_components/adaptive_lighting/switch.py [hunk 8]
```diff
@@ -238,57 +237,136 @@ def _split_service_data(service_data, adapt_brightness, adapt_color):
...
-                light,
-                data[CONF_TRANSITION],
-                data[ATTR_ADAPT_BRIGHTNESS],
-                data[ATTR_ADAPT_COLOR],
-                data[CONF_PREFER_RGB_COLOR],
-                force=True,
-                context=switch.create_context("service", parent=service_call.context),
+    data = service_call.data
+    lights = data[CONF_LIGHTS]
+    switch_unique_id = data.get("entity_id")
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:247]: Is `switch_unique_id` a list of IDs here?

File: custom_components/adaptive_lighting/switch.py [hunk 9]
```diff
@@ -238,57 +237,136 @@ def _split_service_data(service_data, adapt_brightness, adapt_color):
...
+            switch.append(hass.data[DOMAIN][config_id]["instance"])
+    elif lights:
+        # all_adapt_switches = integration_entities(hass, DOMAIN)
+        # No need to check if light is found in multiple switches, that's a user error.
+        config_entries = hass.config_entries.async_entries(DOMAIN)
+        found = False
+        for config in config_entries:
+            if found:
+                break
+            _LOGGER.debug("entry_id: %s", config.entry_id)
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:299]: Some of these `_LOGGER.debug` statements seem a little uninformative for debugging in the future. I would say, either remove them, or make them a bit more informative.

File: custom_components/adaptive_lighting/switch.py [hunk 10]
```diff
@@ -238,57 +237,136 @@ def _split_service_data(service_data, adapt_brightness, adapt_color):
...
-                context=switch.create_context("service", parent=service_call.context),
-            )
+    if conf_entry is not None:
+        ent_reg = entity_registry.async_get(hass)
+        entries = entity_registry.async_entries_for_config_entry(ent_reg, conf_entry)
+        return [entry.entity_id for entry in entries]
+
+    # fallback to just returning all entities for a domain
+    # pylint: disable-next=import-outside-toplevel
+    from .entity import entity_sources
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:363]: Should we just import this at the top?

File: custom_components/adaptive_lighting/switch.py [hunk 11]
```diff
@@ -343,33 +429,103 @@ async def async_setup_entry(
...
     )
 
+    @callback
+    async def handle_apply(service_call: ServiceCall):
+        """Handle the entity service apply."""
+        _LOGGER.debug(
+            "Called 'adaptive_lighting.apply' service with '%s'",
+            service_call.data,
+        )
+        these_switches = data = None
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:439]: This assignment seems unnecessary.

File: custom_components/adaptive_lighting/switch.py [hunk 12]
```diff
@@ -343,33 +429,103 @@ async def async_setup_entry(
...
+                    )
+
+    @callback
+    async def handle_set_manual_control(service_call: ServiceCall):
+        """Set or unset lights as 'manually controlled'."""
+        _LOGGER.debug(
+            "Called 'adaptive_lighting.set_manual_control' service with '%s'",
+            service_call.data,
+        )
+        these_switches = data = None
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:469]: This assignment seems unnecessary.

**Comment by basnijholt** [comment]
Pushed some small changes and left some comments. I have to go now. Will try to continue later today or tomorrow 😄

File: custom_components/adaptive_lighting/switch.py (hunk 10 above)
Comment (th3w1zard1) [reply custom_components/adaptive_lighting/switch.py]: see https://github.com/basnijholt/adaptive-lighting/pull/459/commits/aeed1da0b561e3c2b0aead1d02aab90f930afb7f

File: custom_components/adaptive_lighting/switch.py (hunk 9 above)
Comment (th3w1zard1) [reply custom_components/adaptive_lighting/switch.py]: Looks like you've already done so.

File: custom_components/adaptive_lighting/switch.py (hunk 8 above)
Comment (th3w1zard1) [reply custom_components/adaptive_lighting/switch.py:249]: yes it is a list: a list of switch unique ids. Validates with `cv.entity_ids`.

During testing, I believe `switch_unique_id` started off as a switch object, then a string, then became a dictionary structured as `{<switch unique id>: <switch object>}`, then ended up as a list. That variable has had quite the adventure lol.

File: custom_components/adaptive_lighting/switch.py (hunk 6 above)
Comment (th3w1zard1) [reply custom_components/adaptive_lighting/switch.py]: It's resolved, that's an old reminder to myself during tests when the sequence existed as:

```
        ent_reg = entity_registry.async_get(hass)
        ent_entry = ent_reg.async_get(str(data['entity_id']))
        config_id = ent_entry.config_entry_id
        switch = hass.data[DOMAIN][config_id]["instance"]
```

File: custom_components/adaptive_lighting/switch.py (hunk 7 above)
Comment (th3w1zard1) [reply custom_components/adaptive_lighting/switch.py]: see https://github.com/basnijholt/adaptive-lighting/pull/459/commits/aeed1da0b561e3c2b0aead1d02aab90f930afb7f

File: custom_components/adaptive_lighting/switch.py (hunk 6 above)
Comment (th3w1zard1) [reply custom_components/adaptive_lighting/switch.py]: see https://github.com/basnijholt/adaptive-lighting/pull/459/commits/aeed1da0b561e3c2b0aead1d02aab90f930afb7f

File: custom_components/adaptive_lighting/switch.py (hunk 12 above)
Comment (th3w1zard1) [reply custom_components/adaptive_lighting/switch.py:436]: It needs to initialize somehow.

File: custom_components/adaptive_lighting/switch.py (hunk 11 above)
Comment (th3w1zard1) [reply custom_components/adaptive_lighting/switch.py:406]: It needs to initialize somehow.

Comment (basnijholt) [reply custom_components/adaptive_lighting/switch.py:406]: But you are literally assigning them in the very next line 😅😜

File: custom_components/adaptive_lighting/switch.py [hunk 13]
```diff
@@ -343,33 +429,103 @@ async def async_setup_entry(
...
+    hass.services.async_register(
+        domain=DOMAIN,
+        service=SERVICE_APPLY,
+        service_func=handle_apply,
+        schema=vol.Schema(
+            {
+                vol.Optional("entity_id"): cv.entity_ids,
+                vol.Optional(
+                    CONF_LIGHTS, default=[]
+                ): cv.entity_ids,  # pylint: disable=protected-access
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:471]: Note to self: check whether pylint comment is needed

Comment (th3w1zard1) [reply custom_components/adaptive_lighting/switch.py:471]: It's not, the only time I ever needed it was when I worked on the `change_sunlight_settings` PR which modified the frozen SunLightSettings instance directly.

File: custom_components/adaptive_lighting/switch.py (hunk 11 above)
Comment (th3w1zard1) [reply custom_components/adaptive_lighting/switch.py:406]: Wow so I am. I guess I can't keep up with the amount of code I've moved around.

That can be removed.

**Comment by basnijholt** [comment]
I tried making the title a little bit more descriptive. Please change it as you see fit.

**Comment by th3w1zard1** [comment]
@basnijholt Great! Could you perhaps glance over my comment in #450?

File: custom_components/adaptive_lighting/switch.py [hunk 14]
```diff
@@ -238,57 +237,102 @@ def _split_service_data(service_data, adapt_brightness, adapt_color):
...
+                "_parse_service_args: switches: %s, ent_reg: %s, ent_entry: %s, config_id: %s",
+                switches,
+                ent_reg,
+                ent_entry,
+                config_id,
+            )
+            switches.append(hass.data[DOMAIN][config_id]["instance"])
+        return switches
+    elif lights:
+        switch = _find_switch_with_lights(hass, lights, service_call)
```
Comment (basnijholt) [inline custom_components/adaptive_lighting/switch.py:327]: I have factored the code in the `_find_switch_with_lights` function. However, I believe there exists a bug in code. Previously `switch` could be `None`, `list[AdaptiveSwitch]` and `AdaptiveSwitch`. I believe elsewhere it always assumes a list of switches is returned.

**Comment by th3w1zard1** [comment]
@basnijholt Sorry for the mess! Are there any other PRs you believe I could clean up prior to review?

**Comment by basnijholt** [comment]
@all-contributors please add @firstof9 for code

**Comment by basnijholt** [comment]
@th3w1zard1, no worries at all! I don't mind making a few small changes. Hope you agreed with them 😄

**Comment by basnijholt** [comment]
I also released 1.6.0 just now. So if someone reports issues, we know where they come from :)

**Comment by th3w1zard1** [comment]
I enjoyed watching the commits roll out, the formatting is so clean and I can tell this was done right. I'm surprised at how fast you work. You even added that user error check I didn't bother to add, for when users have lights existing in multiple configs! I even left a reminder to myself 'don't bother checking if the light exists in multiple configs - that's a user error.'

Nice work!

**Pull request merged**